### PR TITLE
Validate cookie name prefixes

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Validates cookie names with prefixes as per https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-05#section-4.1.3.

I'm also not really sure why the other name/value validation raises an `IO::Error`?  Seems more appropriate as an `ArgumentError`.

### CODE REVIEW

File: src/http/cookie.cr [hunk 1]
```diff
@@ -58,6 +60,14 @@ module HTTP
           raise IO::Error.new("Invalid cookie name")
         end
       end
+
+      # Enforce cookie prefix requirements as per https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-05#section-4.1.3.
+      case name
+      when .starts_with? "__Secure-"
+        raise ArgumentError.new "Invalid cookie name.  Has '__Secure-' prefix, but is not secure." unless @secure
```
Comment (Sija) [inline src/http/cookie.cr:67]:
```suggestion
        raise ArgumentError.new "Invalid cookie name. Has '__Secure-' prefix, but is not secure." unless @secure
```

File: spec/std/http/cookie_spec.cr [hunk 2]
```diff
@@ -35,16 +35,27 @@ module HTTP
         expect_raises IO::Error, "Invalid cookie name" do
           HTTP::Cookie.new("", "")
         end
+
```
Comment (straight-shoota) [inline spec/std/http/cookie_spec.cr:38]: What's the reason for these unrelated whitespace changes?

File: src/http/cookie.cr [hunk 3]
```diff
@@ -58,6 +60,14 @@ module HTTP
...
+      when .starts_with? "__Host-"
+        raise ArgumentError.new "Invalid cookie name.  Does not meet '__Host-' prefix requirements." if !@secure || @path != "/" || !@domain.nil?
```
Comment (Sija) [inline src/http/cookie.cr:69]:
```suggestion
        raise ArgumentError.new "Invalid cookie name. Does not meet '__Host-' prefix requirements." if !@secure || @path != "/" || !@domain.nil?
```

**Review by Sija — COMMENTED** [review]
Pedantic: whitespaces

File: src/http/cookie.cr [hunk 4]
```diff
@@ -30,6 +30,7 @@ module HTTP
     # Creates a new `Cookie` instance.
     #
     # Raises `IO::Error` if *name* or *value* are invalid as per [RFC 6265 §4.1.1](https://tools.ietf.org/html/rfc6265#section-4.1.1).
+    # Raises an `ArgumentError` if *name* has a security prefix but is invalid as per [RFC 6265 bis §4.1.3](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-07#section-4.1.3).
```
Comment (straight-shoota) [inline src/http/cookie.cr:33]: With this phrasing the invalidity grammatically refers to the name, instead of the whole cookie.

Maybe put it like this?

```suggestion
    # Raises `ArgumentError` if *name* has a security prefix but the requirements are not met as per [RFC 6265 bis §4.1.3](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-07#section-4.1.3).
```

**Review by straight-shoota — CHANGES_REQUESTED** [review]
I'm doubting this is a good change in terms of API design. Now `validate_name` not just validates the name itself but in the context of other values. This creates some implicit restrictions when using the API in a builder flow, enforcing a specific order of property assignments.

I would suggest to add a dedicated method to validate the entire cookie. This validation could be applied automatically in `parse_set_cookie`.

File: spec/std/http/cookie_spec.cr (hunk 2 above)
Comment (Blacksmoke16) [reply spec/std/http/cookie_spec.cr:38]: No real reason other than it looked bad.

Ref: https://github.com/crystal-lang/crystal/issues/9804.

Comment (straight-shoota) [reply spec/std/http/cookie_spec.cr:38]: Then it probably shouldn't be part of this PR.

**Comment by Blacksmoke16** [comment]
> I would suggest to add a dedicated method to validate the entire cookie. This validation could be applied automatically in `parse_set_cookie`.

@straight-shoota `parse_set_cookie` wouldn't handle all the ways you could create a cookie, e.g. when creating it via `.new` or using the name setter.

> This creates some implicit restrictions when using the API in a builder flow, enforcing a specific order of property assignments.

I don't see how this could be avoided.  Otherwise you could "build" an invalid cookie and wouldn't even know it.

**Comment by straight-shoota** [comment]
The answer is that in order to validate the cookie value, you need to validate explicitly when building an instance. I don't think a setter should apply implicit validation on an incomplete cookie.

**Comment by Blacksmoke16** [comment]
@straight-shoota Oh I see.  You were proposing this method be public, then just also call it in the constructor.  What are your thoughts on how it should work then?  Like `.valid? : Bool`, or `.validate! : Nil | NoReturn`, or?

**Comment by straight-shoota** [comment]
Prob. both?

File: spec/std/http/cookie_spec.cr [hunk 5]
```diff
@@ -84,45 +86,20 @@ module HTTP
...
 
-        expect_raises ArgumentError, "Invalid cookie name.  Has '__Secure-' prefix, but is not secure." do
-          cookie.name = "__Secure-x"
-        end
-
-        cookie.secure = true
-
         cookie.name = "__Secure-x"
         cookie.name.should eq "__Secure-x"
+        cookie.secure.should be_false
```
Comment (straight-shoota) [inline spec/std/http/cookie_spec.cr:94]: Just thinking, but what if a `__Secure` prefix automatically sets `secure = true`?

Comment (Blacksmoke16) [reply spec/std/http/cookie_spec.cr:94]: That probably wouldn't be a bad idea.  Could probably set things up like:

* If it has a security prefix
  * and the related properties are `nil`
    * make the cookie compliant with the prefix
  * and one or more of the related properties are manually set
    * call `validate!`
      * possibly after setting the `nil` ones?

Thoughts?

Comment (Blacksmoke16) [reply spec/std/http/cookie_spec.cr:94]: Taking another look at this. The one thing that could be unexpected is that `@secure` is typed as a `Bool` with a default value of `false`. Because of this we don't really know if the value was passed explicitly as `false` or was just left unset.

We could do one of a few things:

1. Let it be and raise an exception if you dont set `secure: true` with a `__Secure-` prefix
2. Automatically set `@secure = true` if the name has the prefix, arguably could lead to some confusion given it would be set as secure while explicitly passing `secure: false`
3. Change the constructor type/default to be `secure : Bool? = nil` and then `@secure = secure || false`. Then we could pick up on `nil` and base the auto set logic on that
4. Or did you just mean in the context of the `name=` setter?

Comment (straight-shoota) [reply spec/std/http/cookie_spec.cr:94]: 1. Is what's currently implemented. It's just a validation, not auto-detection.
2. would be confusing and thus bad.
3. might work
4. No, but that's a good point. The prefix should probably be validated when assigning `name` and `secure`, not just in the constructor. Currently `Cookie.new("foo", "bar").name = "__Secure-foo"` would not raise an error.

An alternative to 3. would be an additional overload where `secure` parameter is ommitted. But I think changing the default value to `nil` for autodetect seems reasonable.

Comment (Blacksmoke16) [reply spec/std/http/cookie_spec.cr:94]: > 4\. The prefix should probably be validated when assigning `name` and `secure`, not just in the constructor.

I think I had this way back when, but we changed it to public `#valid?` and `#validate!` methods because doing that logic in the setter makes it harder to use the builder pattern to construct the cookie. Unless we're okay with needing to make the changes in a specific order. I.e. set `.secure = true` _before_ `.name = "__Secure-foo"`.

File: src/http/cookie.cr [hunk 6]
```diff
@@ -32,16 +32,31 @@ module HTTP
...
+      if @name.starts_with?("__Host-") && @path.nil? && @domain.nil? && secure.nil?
+        @path = "/"
+        secure = true
+      end
+
+      if @name.starts_with?("__Secure-") && secure.nil?
+        secure = true
+      end
+
+      @secure = secure || false
```
Comment (Sija) [inline src/http/cookie.cr:57]:
```suggestion
      @secure ||= false
```

File: spec/std/http/cookie_spec.cr (hunk 5 above)
Comment (straight-shoota) [reply spec/std/http/cookie_spec.cr:99]: I think I'd prefer to keep the API tied down. It's easy to forget a `validate!` call after building the instance. I suppose this would work better with a dedicated builder API such as `Cookie.build |cookie| cookie.name = "__Secure-foo"; ... end` where validity can be ensured at the end of the `build` method.

But that would probably be too much for this.

Wouldn't it be fine to implicitly set `secure = true` when assigning a name with `__Secure` prefix? If you want to go rogue you can still assign `secure = false` afterwards (and forget about `validate!`) but then you're doing that explicitly.

Comment (Blacksmoke16) [reply spec/std/http/cookie_spec.cr:99]: Thinking a bit more extreme, what exactly is the use case for having a constructor for all the properties, while also letting the cookie be mutable via setters? Like is there actually a need for them versus just creating a cookie with the data you expect all at once? A lot of this would be resolved by just deprecating the setters and force everyone to just use the constructor. That would of course require a different issue/discussion tho.

In regards to this ticket, that approach is fine by me but my only concern is since `@secure` isn't nilable, we wouldn't know if we should make it secure or not. I.e. how to handle:

```cr
cookie.secure = false
cookie.name = "__Secure-foo"
cook.secure # => true
```

Could just be a bit confusing to explicitly set it to `false` but end up getting `true`, depending on the order you do things in. Thoughts on internally making `@secure` nilable? Can keep the API the same by having a custom getter method that would do `@secure || false`. This way we'd at least be able to pick up we shouldn't alter `@secure` since it was explicitly set at some point.

A similar thing is already possible for `__Host-` given `@path` and `@domain` are nilable already.

Comment (straight-shoota) [reply spec/std/http/cookie_spec.cr:99]: Yeah, removing the setters would be a huge change. I assume they exist in order to use `Cookie` with a builder-like API.

Setting `@secure` to `nil` seems like a good idea. Although I probably wouldn't mind `#name=` to override any previous `secure` setting either.

File: src/http/cookie.cr [hunk 7]
```diff
@@ -140,6 +155,44 @@ module HTTP
...
+      raise ArgumentError.new "Invalid cookie name. Has '__Secure-' prefix, but is not secure." unless self.valid_secure_prefix?
+      raise ArgumentError.new "Invalid cookie name. Does not meet '__Host-' prefix requirements." unless self.valid_host_prefix?
+
+      self
+    end
+
+    private def valid_secure_prefix? : Bool
+      has_secure_prefix = @name.starts_with?("__Secure-")
+
+      !has_secure_prefix || (has_secure_prefix && self.secure)
```
Comment (straight-shoota) [inline src/http/cookie.cr:176]: This logic expression can be simplified, which allows to get rid of the variable. Reversing the order also enables to skip the more expensive string comparison operation.

```suggestion
      self.secure || !@name.starts_with?("__Secure-")
```

File: src/http/cookie.cr [hunk 8]
```diff
@@ -140,6 +155,44 @@ module HTTP
...
+    end
+
+    private def valid_host_prefix? : Bool
+      has_host_prefix = @name.starts_with?("__Host-")
+
+      !has_host_prefix || (has_host_prefix && self.secure && "/" == @path && @domain.nil?)
```
Comment (straight-shoota) [inline src/http/cookie.cr:182]: Ditto:

```suggestion
      !@name.starts_with?("__Host-") || (self.secure && "/" == @path && @domain.nil?)
```

File: spec/std/http/cookie_spec.cr [hunk 9]
```diff
@@ -45,6 +45,29 @@ module HTTP
...
       end
+
+      describe "with a security prefix" do
+        it "raises on invalid cookie with prefix" do
+          expect_raises ArgumentError, "Invalid cookie name. Has '__Secure-' prefix, but is not secure." do
+            HTTP::Cookie.new("__Secure-foo", "bar", secure: false)
+          end
+
+          expect_raises ArgumentError, "Invalid cookie name. Does not meet '__Host-' prefix requirements." do
+            HTTP::Cookie.new("__Host-foo", "bar", domain: "foo")
```
Comment (straight-shoota) [inline spec/std/http/cookie_spec.cr:56]: It would be helpful to name the prefix requirements explicitly: secure, no domain, path `/` (don't necessarily need to point out which one exactly failed, but that would be a nice enhancement).

File: spec/std/http/cookie_spec.cr [hunk 10]
```diff
@@ -105,6 +162,59 @@ module HTTP
...
+        cookie.valid?.should be_false
+
+        expect_raises ArgumentError, "Invalid cookie name. Has '__Secure-' prefix, but is not secure." do
+          cookie.validate!
+        end
+
+        cookie.secure = true
+
+        cookie.name = "__Secure-x"
+        cookie.name.should eq "__Secure-x"
```
Comment (straight-shoota) [inline spec/std/http/cookie_spec.cr:180]: Why set `name` again?
```suggestion
```

File: src/http/cookie.cr [hunk 11]
```diff
@@ -20,28 +20,41 @@ module HTTP
...
       @name = name
       validate_value(value)
       @value = value
       raise IO::Error.new("Invalid max_age") if @max_age.try { |max_age| max_age < Time::Span.zero }
+
+      self.check_prefix
+      self.validate!
+    end
+
+    def secure : Bool
```
Comment (straight-shoota) [inline src/http/cookie.cr:53]:
```suggestion
    # Returns `true` if this cookie has the *Secure* flag.
    # The *Secure* flag is implicitly set if `name` has the prefix `__Secure-` or `__Host-`, unless explicitly disabled with `cookie.secure = false`.
    def secure : Bool
```

File: src/http/cookie.cr [hunk 12]
```diff
@@ -140,6 +155,44 @@ module HTTP
...
+    private def valid_host_prefix? : Bool
+      has_host_prefix = @name.starts_with?("__Host-")
+
+      !has_host_prefix || (has_host_prefix && self.secure && "/" == @path && @domain.nil?)
+    end
+
+    private def check_prefix : Nil
+      if @name.starts_with?("__Host-") && @path.nil? && @domain.nil? && @secure.nil?
+        @path = "/"
+        @secure = true
```
Comment (straight-shoota) [inline src/http/cookie.cr:188]: Does the assignment of default values need to be dependent between `path` and `secure`?
I think they should be assignable independent of each other:

```suggestion
      if @name.starts_with?("__Host-")
        @path = "/" if @path.nil?
        @secure = true if @secure.nil?
```

File: src/http/cookie.cr [hunk 13]
```diff
@@ -140,6 +155,44 @@ module HTTP
...
+      end
+
+      if @name.starts_with?("__Secure-") && @secure.nil?
+        @secure = true
```
Comment (straight-shoota) [inline src/http/cookie.cr:192]: For maybe a bit more clarity (and congruence with the previous suggestion):
```suggestion
      if @name.starts_with?("__Secure-")
        @secure = true if @secure.nil?
```

File: src/http/cookie.cr [hunk 14]
```diff
@@ -173,13 +173,24 @@ module HTTP
     private def valid_secure_prefix? : Bool
       has_secure_prefix = @name.starts_with?("__Secure-")
 
-      !has_secure_prefix || (has_secure_prefix && @secure)
+      !has_secure_prefix || (has_secure_prefix && self.secure)
```
Comment (Sija) [inline src/http/cookie.cr:176]:
```suggestion
      !has_secure_prefix || (has_secure_prefix && secure)
```

File: src/http/cookie.cr [hunk 15]
```diff
@@ -173,13 +173,24 @@ module HTTP
...
     end
 
     private def valid_host_prefix? : Bool
       has_host_prefix = @name.starts_with?("__Host-")
 
-      !has_host_prefix || (has_host_prefix && @secure && "/" == @path && @domain.nil?)
+      !has_host_prefix || (has_host_prefix && self.secure && "/" == @path && @domain.nil?)
```
Comment (Sija) [inline src/http/cookie.cr:182]:
```suggestion
      !has_host_prefix || (has_host_prefix && secure && "/" == @path && @domain.nil?)
```

File: src/http/cookie.cr (hunk 12 above)
Comment (Blacksmoke16) [reply src/http/cookie.cr:188]: This is expected in this context since `__Host-` _requires_ it to be secure, have no domain, and a path of `"/"`. 

Ref: https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-05#section-4-1-3-2

**Review by straight-shoota — APPROVED** [review]

File: src/http/cookie.cr [hunk 16]
```diff
@@ -140,6 +156,40 @@ module HTTP
...
+    # Raises `ArgumentError` if `self` is not `#valid?`.
+    def validate! : self
+      raise ArgumentError.new "Invalid cookie name. Has '__Secure-' prefix, but is not secure." unless self.valid_secure_prefix?
+      raise ArgumentError.new "Invalid cookie name. Does not meet '__Host-' prefix requirements of: secure: true, path: \"/\", domain: nil." unless self.valid_host_prefix?
+
+      self
+    end
+
+    private def valid_secure_prefix? : Bool
+      self.secure || !@name.starts_with?("__Secure-")
```
Comment (Sija) [inline src/http/cookie.cr:175]:
```suggestion
      self.secure || (@secure.nil? && !@name.starts_with?("__Secure-"))
```

Comment (straight-shoota) [reply src/http/cookie.cr:175]: Hm, yeah but shouldn't it be `@secure != false || !@name.starts_with?("__Secure-")`?

Comment (Blacksmoke16) [reply src/http/cookie.cr:175]: Is this scenario even possible? I'm struggling to think about how it would get into that state.

* Can't pass `nil` to `secure=`
* If you use an explicit `nil` in the constructor w/o the prefix, its valid
* If you use an explicit `nil` in the constructor with the prefix, it gets set to `true`
* If you set the name after creating it with no or `nil` value, it still gets set the `true`

Comment (straight-shoota) [reply src/http/cookie.cr:175]: Yeah, I can't think of how to reach that state. But I suppose it still wouldn't hurt to cover this case.

**Review by straight-shoota — APPROVED** [review]

File: src/http/cookie.cr (hunk 16 above)
Comment (Sija) [reply src/http/cookie.cr]: > Hm, yeah but shouldn't it be `@secure != false || !@name.starts_with?("__Secure-")`?

@straight-shoota I don't think so, since `nil != false # => true`

Comment (Blacksmoke16) [reply src/http/cookie.cr]: Mmm good call. I did think having both `be_true` in the next test case was kinda weird. I went back and played with it more. I think it should actually be `self.secure || !@name.starts_with?("__Secure-")`. This makes all tests pass, after changing the one to `be_false` after setting `@secure` to `nil`.

The cookie would always be valid if it's true, otherwise it's only valid if it doesnt have the prefix.

Comment (Sija) [reply src/http/cookie.cr:175]: The most correct is IMO `self.secure || (@secure.nil? && !@name.starts_with?("__Secure-"))`.

Comment (Blacksmoke16) [reply src/http/cookie.cr:175]: That causes a bunch of spec failures, such as causing `HTTP::Cookie.new "x", "", secure: false` to raise saying it has the prefix but isn't secure. Which isn't correct.

**Review by sdogruyol — APPROVED** [review]
Thank you @Blacksmoke16  🙏
